src/closer.rs

```diff
@@ -0,0 +1,58 @@
+use std::sync::{Arc, Mutex};
+
+use crossbeam_channel::{unbounded, Receiver, Sender};
+
+type CloserChanType = Arc<(Mutex<Option<Sender<()>>>, Receiver<()>)>;
+
+// TODO: review closer implementation
+#[derive(Clone)]
+pub struct Closer {
+    chan: CloserChanType,
+}
+
+impl Closer {
+    pub fn new() -> Self {
+        let (tx, rx) = unbounded::<()>();
+        Self {
+            chan: Arc::new((Mutex::new(Some(tx)), rx)),
+        }
+    }
+
+    pub fn close(&self) {
+        self.chan.0.lock().unwrap().take();
+    }
+
+    pub fn has_been_closed(&self) -> &Receiver<()> {
```

Comment:
The name "has_been_closed" typically returns bool, to return a Receiver, why not call get_receiver?